framework/src/vectorpostprocessors/ElementMaterialSampler.C

```diff
@@ -44,35 +46,62 @@ ElementMaterialSampler::ElementMaterialSampler(const InputParameters & parameter
     _y_coords(declareVector("y")),
     _z_coords(declareVector("z"))
 {
-  auto & mat = getMaterialByName(getParam<MaterialName>("material"), true);
-  auto & prop_names = mat.getSuppliedItems();
-  if (mat.isBoundaryMaterial())
-    mooseError(name(), ": boundary materials (i.e. ", mat.name(), ") cannot be used");
+  // Check either "material" or "property" was set but not both
+  if (parameters.isParamSetByUser("material") && parameters.isParamSetByUser("property"))
+    mooseError("Setting both 'material' and 'property' is not allowed. Use one or the other.");
+  if (!parameters.isParamSetByUser("material") && !parameters.isParamSetByUser("property"))
+    mooseError("Either 'material' and 'property' needs to be set.");
+
+  // List of property names to collect
+  std::vector<MaterialName> prop_names;
 
   // Get list of elements from user
   if (parameters.isParamValid("elem_ids"))
   {
     const auto & ids = getParam<std::vector<dof_id_type>>("elem_ids");
     _elem_filter = std::set<dof_id_type>(ids.begin(), ids.end());
+  }
+
+  // If Material is used, get all properties.
+  if (parameters.isParamSetByUser("material"))
+  {
+    auto & mat = getMaterialByName(getParam<MaterialName>("material"), true);
+    if (mat.isBoundaryMaterial())
+      mooseError(name(), ": boundary materials (i.e. ", mat.name(), ") cannot be used");
 
-    // check requested materials are available
-    for (const auto & id : ids)
+    // Get property names from the Material
+    auto & props = mat.getSuppliedItems(); // returns std::set
+    prop_names = std::vector<MaterialName>(props.begin(), props.end());
+
+    // Check requested materials are available
+    if (_elem_filter)
```

Comment:
I would move this check down a bit for clarity. 

And in the case where we only know the property, you can just poll every single material in the problem to find if it defines a property. If you do it once before the element loop it should be fine. 

It could be good also to turns the element loop into a simple gathering of the block ids. Then once you have those in a set you can test the materials outside of an element loop

Reply:
> poll every single material in the problem to find if it defines a property

How do you do that?

> gathering of the block ids

Not sure how to do that either.

Reply:
this check still needs to be used in both cases. Maybe dont share the code, just have two different checks for the material and material property cases

instead of that
```
And in the case where we only know the property, you can just poll every single material in the problem to find if it defines a property. If you do it once before the element loop it should be fine.
```
use this, that should be a lot easier
```
  /**
   * Get a vector containing the block ids the material property is defined on.
   */
  virtual std::set<SubdomainID> getMaterialPropertyBlocks(const std::string & prop_name);
```

Reply:
nevermind. This is covered